Clean up unsafety in char::encode_utf8

Mark-Simulacrum opened the pull request:
This originally started as an attempt to allow LLVM to optimize through
encode_utf8 to detect the try_encode_utf8 case (#52579, #52580), but due to a
typo my conclusion that my optimizations were successful was incorrect.

Furthermore, as far as I can tell, this optimization is probably just not
possible with LLVM today.  This [code](https://rust.godbolt.org/z/JggRj4)
compiles down to a long series of compares, notably, two identical series of
compares. That essentially means that LLVM is today unable to see that these two
ifs are identical and as such can be merged and then realize that no value of
the if condition can result in a call to `please_delete`. As such, for now, we
do not attempt to specifically optimize for that case.

rust-highfive commented:
r? @cramertj

(rust_highfive has picked a reviewer for you, use r? to override)

Mark-Simulacrum commented:
See also https://github.com/rust-lang/rust/pull/52580 for some context, where @SimonSapin and @oli-obk [suggested](https://github.com/rust-lang/rust/pull/52580#issuecomment-411516054) that instead of defining the non-panicking version of the function we would rather see the "obvious" bounds check optimize well.

oli-obk commented on src/libcore/char/methods.rs:
```diff
@@ -455,9 +455,13 @@ impl char {
                 4
             } else {
+                // We know that all chars fit inside 4 bytes
+                if dst.len() <= 4 {
```
Shouldn't this be `< 4`?

ranma42 replied:
or rather `>= 4`
example: if `self` requires more than one byte to be encoded and `dst.len` is 1, this code is reached and should not be considered unreachable.

oli-obk replied:
In the old code it wasn't reachable. After the slice changes that came after my comment I need to reevaluate.

Mark-Simulacrum replied:
I'm re-verifying... I think @ranma42 might be right, but in any case, I think I found an LLVM bug -- https://rust.godbolt.org/z/ddVHt8 should optimize out, i.e., get an empty `bar` function and so on, since there's no way for the panic to be hit, but doesn't. That might make getting the optimization to fire here quite difficult.

I'm attempting now to figure out whether we can do something in Rust to convince LLVM.

Mark-Simulacrum commented:
r? @oli-obk

I updated the PR description, but to describe the changes since the last version (rather than the changes vs. master):

I've further removed almost all unsafe code from the function now (just the cast to `str` is still left as unsafe). I've also confirmed that nightly and this branch produce essentially identical assembly for the following code. I've given up on trying to resolve https://github.com/rust-lang/rust/issues/52579 in this PR per the PR description's note that there is just an LLVM bug here that I have no reason to believe we can work around currently.

```rust
pub fn try_encode_utf8(x: char, buf: &mut [u8]) -> Option<&mut str> {
    if x.len_utf8() <= buf.len() {
        Some(x.encode_utf8(&mut buf[..]))
    } else {
        None
    }
}

pub fn encode_utf8_4(x: char, buf: &mut [u8; 4]) -> &mut str {
    x.encode_utf8(&mut buf[..])
}
```

oli-obk commented on src/libcore/char/methods.rs:
```diff
@@ -434,36 +434,35 @@ impl char {
+                dst.len(),
+            ),
+        };
+        // SAFETY: We also just wrote UTF-8 content in, so converting to str is fine.
```
```suggestion
        // SAFETY: We just wrote UTF-8 content in, so converting to str is fine.
```

There's no other thing that the `also` refers to

oli-obk commented:
Maybe we could come up with a general MIR optimization that can eliminate the branch already at MIR time.

r=me with comment nit fixed

Mark-Simulacrum commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit df4d490038c37e441065890fa27ed2ce0bdf83e6 has been approved by `oli-obk`

<!-- @bors r=oli-obk df4d490038c37e441065890fa27ed2ce0bdf83e6 -->
<!-- homu: {"type":"Approved","sha":"df4d490038c37e441065890fa27ed2ce0bdf83e6","approver":"oli-obk"} -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 100, this pull request will be tested once the tree is reopened
